### ISSUE TITLE
COOL Writer: Compare two files programmatically with uno:CompareDocuments

### ISSUE THREAD

**Issue by jaime-rivas** [CollaboraOnline/online#4535]
Hi everyone!

It would be much appreciated to add the uno:CompareDocuments command to programmatically compare two files with Writer on the server. Today, users have to manually open Writer in their local computer, select the two files to compare, and then see the compared file. It would a fantastic addition to the Collabora Online API to have a way to programmatically make that comparison on the server and return the compared file to the user.

Thank you so much for the great work on Collabora Online!

**Comment by Andreas-Kainz** [comment]
Yes uno:CompareDocuments will be very useful at least when you use cool within a cloud storage where you have for example a version history. Compare the actual "master" document with an older version would be awesome. @juliushaertl

**Comment by pedropintosilva** [comment]
oh I forget to link here what originated this feature request: https://forum.collaboraonline.com/t/cool-writer-compare-two-files-files-programmatically 

Yes @Andreas-Kainz but as I stated there it is not so easy and it would involved further work on the Online side and how to manage 3 different views within a user etc etc

**Comment by Andreas-Kainz** [comment]
Do we need 3 views or is it enough to show the main view with the comments and changes?

**Comment by jaime-rivas** [comment]
I think the main view with the comments / changes next to it, would be fantastic (screenshot 1):

![Writer compare](https://user-images.githubusercontent.com/19812916/163905023-74c43508-a18c-4946-8803-4f5398fc443e.png)

Also, if the API could accept user preferences for colors, attributes (the options from screenshot 2), then it would be just perfect:

![Changes options, screenshot 2](https://user-images.githubusercontent.com/19812916/163905061-cec7cca0-58dc-495e-a6c5-2ec080b7e63b.png)

Thank you so much.

**Comment by pedropintosilva** [comment]
Thank you @jaime-rivas for the valuable input :)

**Comment by juliushaertl** [comment]
For the Nextcloud integration the WOPI endpoints already support fetching different versions of the file, so once there is a way to open the version diff for two provided WOPI urls integration on our side would be relatively straight forward.

**Comment by Andreas-Kainz** [comment]
> I think the main view with the comments / changes next to it, would be fantastic (screenshot 1):
> 
> ![Writer compare](https://user-images.githubusercontent.com/19812916/163905023-74c43508-a18c-4946-8803-4f5398fc443e.png)

How would it look like when you open the diff file from LibO within cool? Cause this view you should see also in cool isn't it?

**Comment by pedropintosilva** [comment]
> manage 3 different views within a user etc etc

actually I was confused I remember to discuss something with @mmeeks and I think I somehow though about something more complex. So I'm wrong we could just reuse tracking changes that we already have in place but of course some more work needs to be done but definitely  not the amount I was mentioning

**Comment by jancborchardt** [comment]
Good stuff! General question, as this is yet another place of UI elements (next to the right sidebar and the comments bar): Wouldn’t this make sense in the right sidebar as well when you are reviewing the changes?

**Comment by jaime-rivas** [comment]
@jancborchardt Yes, the the right sidebar seems like the perfect place for the list of changes:

![image](https://user-images.githubusercontent.com/19812916/164012226-63c8afe8-9c1f-496a-bcf3-b8b26a7406e0.png)

**Comment by pedropintosilva** [comment]
check > is it ok, to show only the changed document and not all 3 views

**Comment by pedropintosilva** [comment]
right now we already have the Manage changes available on online but as a modal dialog: 
![image](https://user-images.githubusercontent.com/65948705/164024587-0c2ee34e-0417-46d9-9baa-6984d48e3ed1.png)

**Comment by pedropintosilva** [comment]
@Andreas-Kainz : yup I think it makes sense, so it is within the same places (both on classic and nb) as tracking changes' actions

**Comment by jaime-rivas** [comment]
In addition to having uno:CompareDocuments in the Tools menu and in the Review tab within Collabora, I think it would be ideal if we can send the reference of the two files to the API (probably from a dialog outside of Collabora, from the user's file server service), and then it opens directly Collabora Online in "Compared view". Something similar to these two examples:
https://docs.microsoft.com/en-us/office/vba/api/word.document.compare
https://blog.aspose.com/2020/10/05/compare-ms-word-documents-using-java/

Also, to follow-up on @Andreas-Kainz comment:

- [ ] can we show the same colors than we have in LibO (red, blue, green): yeah it would be ideal to have blue for insertions, red for deletions and green for moves.

- [ ] managed changes within a sidebar and not a dialog: I understand this is a yes

Very excited for this feature!

**Comment by Andreas-Kainz** [comment]
Hi,

The main issue is, that's it is very difficult to bring the compare feature 100% to cool. Cause cool show in general only one content stream and the compare feature show more views.

I can't help with any dev stuff. From my point of view the best would be:
- show the compared final view so you have to run only the compare command in libo and than have the differences as changes.
- for the changes I would like to see an sidebar where you can see all the changes and maybe offers different other stuff.
- than we can talk about how to show the different views.


Anyhow it look like more work, so if you want help best would be to make a research how other online office suites did it, how the GitHub diff view work, how office suits did it and what would work best for online at desktop and mobile devices.

**Comment by jaime-rivas** [comment]
I think that just having the final view of the compared document would be a fantastic first step, without showing the changes panel or anything else in the UI. Once the uno:CompareDocuments API is added to Collabora Online, we can contribute directly to the web viewer to help with the views. 

Does it sound good to first add the API to send two documents to compare and get the final compared file, and then we iterate from there?

**Comment by mmeeks** [comment]
Its easy enough to add those APIs of course - the uno:CompareDocuments command is also reasonably easy to enable in the backend - just git grep doc_iniUnoCommands # in core.git - but this of course is going to want a file-picker to select another document to compare with and so on: more of a pain. I liked your website and UI video for change-tracking https://synergy.page/ that allows the user to select which document to use as the base for comparison =) [ given that these things are never perfect ]. Also gave you a connect on LinkedIn.

**Comment by vmiklos** [comment]
I'll look at this. Note that this issue is specifically about an API to compare two documents, not the wider "visual compare of two documents", which is related, but should be a separate issue.

**Comment by vmiklos** [comment]
https://gerrit.libreoffice.org/c/core/+/194237 and https://github.com/CollaboraOnline/online/pull/13538 implement this, the second is now pending CI, then review.

**Comment by vmiklos** [comment]
The above is now in, should be in the next 25.04.8.1 release. Sample curl invocation:

```
curl -k -F "data=@/path/to/new.docx" -F "compare=@/path/to/old.docx" -F "format=pdf" -o out.pdf https://localhost:9980/cool/convert-to
```

Sample input: [old.docx](https://github.com/user-attachments/files/23678616/old.docx) and [new.docx](https://github.com/user-attachments/files/23678618/new.docx).

The https://sdk.collaboraonline.com/docs/conversion_api.html page will be updated once the release is out.

**Comment by mmeeks** [comment]
Great work - thanks Miklos :-)

**Comment by rayannm** [comment]
anyone know how long it will take for the release of v25.04.8.1 ? Thanks :)

**Comment by jaime-rivas** [comment]
This is great, thank you @vmiklos!

One follow-up question: is it possible to customize the colors used to display changes? In the UI, users can choose colors for additions, moves, and deletions. Can this also be configured via the API? Or, alternatively, is there a way to define these colors once so that the same scheme is always used?

**Comment by vmiklos** [comment]
> anyone know how long it will take for the release of v25.04.8.1 ?

The git tag co-25.04.7-1 was tagged on 4th Nov, 25.04.8.1 should be there roughly 1 month later.

> One follow-up question: is it possible to customize the colors used to display changes?

Not at the moment, but probably it would be useful to file a follow-up github issue for that, should be possible to do so.

**Comment by jaime-rivas** [comment]
Thank you @vmiklos 

As a follow‑up, I’ve opened a new issue to track that enhancement: [Convert-to compare API: customize colors of changes](#13570).
